full_configuration.md
MagiceUserBundle
================

The Magice User Bundle

### Install Instruction

```yaml
# app/config/config.yml
framework:
    translator: ~

doctrine:
    dbal:
        types:
            phone_number: Misd\PhoneNumberBundle\Doctrine\DBAL\Types\PhoneNumberType

    orm:
        auto_generate_proxy_classes: "%kernel.debug%"
        auto_mapping: true
        filters:
            softdeleteable:
                class: Gedmo\SoftDeleteable\Filter\SoftDeleteableFilter
                enabled: true

stof_doctrine_extensions:
    default_locale: "%locale%"
    orm:
        default:
            softdeleteable: true
            timestampable: true

hwi_oauth:
    #this is my custom user provider, created from FOSUBUserProvider - will manage the
    #automatic user registration on your site, with data from the provider (facebook. google, etc.)
    #and also, the connecting part (get the token and the user_id)
    connect:
        account_connector: mg.user.provider
    # name of the firewall in which this bundle is active, this setting MUST be set
    firewall_name: main
    fosub:
        username_iterations: 30
        properties:
            # these properties will be used/redefined later in the custom FOSUBUserProvider service.
            facebook: '--facebook_id--'
            google: '--google_id--'
    resource_owners:
        facebook:
            type:                facebook
            user_response_class: Magice\Bundle\UserBundle\OAuth\Response\Facebook
            client_id:           "xxxx"
            client_secret:       "xxxx"
            scope:               "public_profile,user_friends,email,user_birthday,user_likes,user_location,publish_stream"

fos_user:
    db_driver: orm # other valid values are 'mongodb', 'couchdb' and 'propel'
    firewall_name: main
    user_class: Magice\Bundle\UserBundle\Model\User
    group:
        group_class: Magice\Bundle\UserBundle\Model\Group
    registration:
        form:
            type: mg_user_form_type_registration
            name: mg_user_form_type_registration
    from_email:
        address: noreply@joyprice.com
        sender_name: Joyprice Notify Mailer

twig:
    globals:
        img:
            blank: data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7
```

```yaml
# app/config/routing.yml
mg_user:
    resource: "@MagiceUserBundle/Resources/config/routing.yml"
    prefix:   /user
```

```yaml
# app/config/security.yml
security:
    encoders:
        FOS\UserBundle\Model\UserInterface: sha512

    role_hierarchy:
        ROLE_ADMIN:       ROLE_USER
        ROLE_SUPER_ADMIN: ROLE_ADMIN

    providers:
        fos_userbundle:
            id: fos_user.user_provider.username_email

    firewalls:
        dev:
            pattern:  ^/(_(profiler|wdt)|css|images|js)/
            security: false

        main:
            context:     user
            pattern:     /.*

            form_login:
                provider: fos_userbundle
                csrf_provider: form.csrf_provider
                check_path: /user/login_check
                login_path: /user/login
                default_target_path: /
                use_forward:  false
                use_referer: true

            remember_me:
                key: %secret%
                name: APP_REMEMBER_ME
                lifetime: 31536000
                always_remember_me: true
                remember_me_parameter: _remember_me

            oauth:
                resource_owners:
                    facebook: "/user/login/check-facebook"
                    google: "/user/login/check-google"
                login_path: /user/login
                failure_path: /user/login
                oauth_user_provider:
                    service: mg.user.provider

            logout:
                path: /user/logout
                target: /user/login

            anonymous: true

    access_control:
        #- { path: ^/login, roles: IS_AUTHENTICATED_ANONYMOUSLY, requires_channel: https }
        - { path: ^/user/login$, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/user/register, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/user/resetting, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/user/connect.*, role: IS_AUTHENTICATED_ANONYMOUSLY }

```

```php
# app/AppKernel.php
new Stof\DoctrineExtensionsBundle\StofDoctrineExtensionsBundle(),
new Misd\PhoneNumberBundle\MisdPhoneNumberBundle(),
new HWI\Bundle\OAuthBundle\HWIOAuthBundle(),
new FOS\UserBundle\FOSUserBundle(),
new Magice\Bundle\UserBundle\MagiceUserBundle(),
```




short_configuration.md
### MagiceUserBundle configuration reference.

````yaml
magice_user:
    driver: orm
    path_prefix: /user
    # firewall name (magice build-in firewall)
    firewall: magice
    # User provider service
    provider: mg.user.provider
    username_iterations: 30
    remember_lifetime: 31536000
    
    email:
        address: webmaster@example.com
        sender: webmaster
        # send mail service
        service: mg.user.mailer.twig_swift
    
    form:
        # FOSUser registration form
        registration:
            name: mg_user_form_type_registration
            type: mg_user_form_type_registration
        
    # shorthand facebook config for oauth    
    facebook:
        client_id: xxx
        client_secret: xxx
        scope: xxx
        
    # Full oauth config
    oauth:
        facebook:
            client_id: xxx
            client_secret: xxx
            scope: xxx
            user_response_class: xxx
            service: xxx
            type: xxx
            
    class:
        user: Magice\Bundle\UserBundle\Entity\User
        group: Magice\Bundle\UserBundle\Entity\Group
        
        # oauth respons handle (hwi_oauth bundle)
        responder:
            facebook: Magice\Bundle\UserBundle\OAuth\Response\Facebook
            ....
        
        doctrine:
            # phone_number type for DBAL
            phone_number: Misd\PhoneNumberBundle\Doctrine\DBAL\Types\PhoneNumberType
            # soft delete filter class
            softdeleteable: Gedmo\SoftDeleteable\Filter\SoftDeleteableFilter

````


DependencyInjection/Configuration.php
<?php

namespace Magice\Bundle\UserBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * This is the class that validates and merges configuration from your app/config files
 * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritDoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode    = $treeBuilder->root('magice_user');

        $supportedDrivers = array('orm', 'custom');

        $rootNode
            ->children()
                ->scalarNode('driver')->defaultValue('orm')
                    ->validate()
                        ->ifNotInArray($supportedDrivers)
                        ->thenInvalid('The driver %s is not supported. Please choose one of '.json_encode($supportedDrivers))
                    ->end()
                ->end()

                ->scalarNode('path_prefix')->defaultValue('/user')->end()
                ->scalarNode('firewall')->defaultValue('magice')->end()
                ->scalarNode('firewall_pattern')->defaultValue('/.*')->cannotBeEmpty()->end()
                ->scalarNode('provider')->defaultValue('mg.user.provider')->end()
                ->scalarNode('confirmation')->defaultValue(true)->end()
                ->scalarNode('username_iterations')->defaultValue(30)->end()
                ->scalarNode('remember_lifetime')->defaultValue(31536000)->end()
                ->scalarNode('remember_param_name')->defaultValue('_remember_me')->end()
                ->scalarNode('remember_cookie_name')->defaultValue('MAGICE_REMEMBER_ME')->end()

                ->arrayNode('class')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('user')->defaultValue('Magice\Bundle\UserBundle\DefaultEntity\User')->cannotBeEmpty()->end()
                        ->scalarNode('group')->defaultValue('Magice\Bundle\UserBundle\DefaultEntity\Group')->cannotBeEmpty()->end()
                        ->scalarNode('info')->defaultValue('Magice\Bundle\UserBundle\DefaultEntity\UserInfo')->cannotBeEmpty()->end()
                        ->arrayNode('responder')
                            ->useAttributeAsKey('name')
                            ->prototype('scalar')->cannotBeEmpty()->end()
                        ->end()
                        ->arrayNode('doctrine')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('phone_number')
                                    ->defaultValue('Misd\PhoneNumberBundle\Doctrine\DBAL\Types\PhoneNumberType')
                                    ->cannotBeEmpty()
                                ->end()
                                ->scalarNode('softdeleteable')
                                    ->defaultValue('Gedmo\SoftDeleteable\Filter\SoftDeleteableFilter')
                                    ->cannotBeEmpty()
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()

                ->arrayNode('email')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('address')->defaultValue('webmaster@example.com')->cannotBeEmpty()->end()
                        ->scalarNode('sender')->defaultValue('webmaster')->cannotBeEmpty()->end()
                        ->scalarNode('service')->defaultValue('fos_user.mailer.twig_swift')->cannotBeEmpty()->end()
                    ->end()
                ->end()

                ->arrayNode('form')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->arrayNode('registration')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('name')->defaultValue('mg_user_form_type_registration')->cannotBeEmpty()->end()
                                ->scalarNode('type')->defaultValue('mg_user_form_type_registration')->cannotBeEmpty()->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()

                ->arrayNode('oauth')
                    ->useAttributeAsKey('name')
                    ->prototype('array')
                        ->children()

                            ->scalarNode('client_id')->cannotBeEmpty()->end()
                            ->scalarNode('client_secret')->cannotBeEmpty()->end()

                            ->scalarNode('scope')
                                ->validate()
                                    ->ifTrue(function($v) {
                                            return empty($v);
                                        })
                                    ->thenUnset()
                                ->end()
                            ->end()

                            ->scalarNode('user_response_class')
                                ->validate()
                                    ->ifTrue(function($v) {
                                            return empty($v);
                                        })
                                    ->thenUnset()
                                ->end()
                            ->end()

                            ->scalarNode('service')
                                ->validate()
                                    ->ifTrue(function($v) {
                                            return empty($v);
                                        })
                                    ->thenUnset()
                                ->end()
                            ->end()

                            ->scalarNode('type')
                                ->validate()
                                    ->ifTrue(function($v) {
                                            return empty($v);
                                        })
                                    ->thenUnset()
                                ->end()
                            ->end()

                        ->end()
                    ->end()
                ->end()

            ->end()
        ;

        $this->shorthandOauthFacebook($rootNode);

        return $treeBuilder;
    }

    private function shorthandOauthFacebook(ArrayNodeDefinition $rootNode)
    {
        $rootNode
            ->children()
                ->arrayNode('facebook')
                    ->children()
                        ->scalarNode('client_id')->cannotBeEmpty()->end()
                        ->scalarNode('client_secret')->cannotBeEmpty()->end()

                        ->scalarNode('scope')
                            ->validate()
                                ->ifTrue(function($v) {
                                        return empty($v);
                                    })
                                ->thenUnset()
                            ->end()
                        ->end()

                        ->scalarNode('user_response_class')
                            ->defaultValue('Magice\Bundle\UserBundle\OAuth\Response\Facebook')
                            ->cannotBeEmpty()
                        ->end()

                        ->scalarNode('service')
                            ->validate()
                                ->ifTrue(function($v) {
                                        return empty($v);
                                    })
                                ->thenUnset()
                            ->end()
                        ->end()

                    ->end()
                ->end()
            ->end()
        ;
    }
}
